File: MyApplication/app/src/main/java/io/ckl/boilerplate/modules/main/MainPresenter.java
package io.ckl.boilerplate.modules.main;

import io.ckl.boilerplate.api_services.GreetingAPIService;
import io.ckl.boilerplate.models.Greeting;

/**
 * This is a example presenter.
 *
 * The presenter holds a instance of the View, which is a interface implementation.
 * This view should be set as null whenever the activity reaches onDestroy().
 *
 * The presenter is responsible for the business logic, fetching models and telling the view to update.
 */
public class MainPresenter implements MainInterfaces.Presenter {

    MainInterfaces.View view;

    public MainPresenter(MainInterfaces.View view) {
        this.view = view;
    }

    // region MainInterfaces.Presenter

    @Override
    public void onCreate() {
        greet();
    }

    @Override
    public void onGreetButtonPressed() {
        greet();
    }

    @Override
    public void onDestroy() {
        this.view = null;
    }

    // end region


    // region private

    private void greet() {
        Greeting greeting = GreetingAPIService.fetchGreeting();
        if (view == null) { return; }
        view.showGreeting(greeting.getContent());
    }

    // end region
}

File: MyApplication/app/src/main/java/io/ckl/boilerplate/modules/main/MainPresenter.kt
package io.ckl.boilerplate.modules.main

import android.app.Activity
import android.content.Intent
import io.ckl.boilerplate.models.Greeting

/**
 * This is a example presenter.

 * The presenter holds a instance of the View, the Interactor and the Router, which are all interfaces implementations.
 * Those should be set as null whenever the activity reaches onDestroy().

 * The presenter is responsible for querying the Interactor for Data and telling the view to update.
 */
class MainPresenter(internal var view: MainContracts.View?): MainContracts.Presenter, MainContracts.InteractorOutput {
    var interactor: MainContracts.Interactor? = MainInteractor(this)
    var router: MainContracts.Router? = MainRouter(view as? Activity)

    // region MainContracts.Presenter

    override fun onCreate(intent: Intent?) {
        interactor?.fetchGreeting()
    }

    override fun onGreetButtonPressed() {
        interactor?.fetchGreeting()
    }

    override fun onDestroy() {
        interactor?.unRegister()
        router?.unRegister()
        view = null
        interactor = null
        router = null
    }

    // end region


    // region MainContracts.InteractorOutput

    override fun onGreetingFetched(greeting: Greeting) {
        view?.showGreeting(greeting.content)
    }

    // end region
}
